### PR TITLE
Re-export more names in `sphinx.domains.python`

### PR DESCRIPTION
**Pull request by JacobChesslo**
Subject: Re-export various names in `sphinx.domains.python`

### Feature or Bugfix
- Bugfix

### Purpose
- Re-export various names in `sphinx.domains.python`

### Detail
- Re-export various names in `sphinx.domains.python`

### Relates
- https://github.com/sphinx-doc/sphinx/issues/12295

### CODE REVIEW

File: sphinx/domains/python/__init__.py [hunk 1]
```diff
@@ -38,6 +38,16 @@
...
 # re-export objects for backwards compatibility
 # xref https://github.com/sphinx-doc/sphinx/issues/12295
+from sphinx.domains.python._annotations import (  # NoQA: F401
+    _parse_annotation,
+    _parse_arglist,
+    _parse_type_list,
+    _pseudo_parse_arglist,
+    _TypeParameterListParser,
+    parse_reftarget,
+    type_to_xref,
```
Comment (AA-Turner) [inline sphinx/domains/python/__init__.py]: I'm uneasy about this -- these are mainly private names, and hence fragile. ``sphinx-immaterial`` should have resilience if the names go away, and ideally not use them at all.

Also note, I am only able to find uses of ``type_to_xref`` and ``_parse_arglist`` in ``sphinx-immaterial``.

Comment (JacobChesslo) [reply sphinx/domains/python/__init__.py]: I don't blame you about being queasy on the more private imports.

I'm currently trying to see how far down these imports go, and if I can get a working install again, then working back from there.

Comment (JacobChesslo) [reply sphinx/domains/python/__init__.py]: I was able to get the originally linked issue's reproducibility steps to again work with these proposed changes, so that's hopeful that a subset of the changes can definitely fully resolve this issue.

I definitely agree on the private imports being more problematic to include in, so I've also created https://github.com/jbms/sphinx-immaterial/issues/345 to potentially work this from both ends.

**Comment by JacobChesslo** [comment]
I think given the initial blast radius of https://github.com/sphinx-doc/sphinx/issues/12295, it may be worth making these private methods more accessible and/or putting them on a more rigorous deprecation schedule.

**Comment by jbms** [comment]
sphinx-immaterial heavily monkey patches sphinx to add functionality and indeed it is on us to deal with changes in sphinx. 
 That said, monkey patching sphinx is not exactly unusual --- and sphinx itself does some monkey patching of docutils.  We would like to integrate this functionality upstream but there are challenges with that.  Exactly what name they have or whether they start with an underscore doesn't really matter but renames do create additional work for us.

**Comment by chrisjsewell** [comment]
> sphinx itself does some monkey patching of docutil

oh trust me, I'm not happy about that either 😄

> monkey patching sphinx is not exactly unusual
> Exactly what name they have or whether they start with an underscore doesn't really matter but renames do create additional work for us.

but yes, whether it usual or not (I wouldn't say so much), its absolutely not within  the "contract" of any package to care or account for this type of usage.
If you do, especially without the knowledge of the upstream package, then you are just creating tech debt for yourself

> We would like to integrate this functionality upstream but there are challenges with that. 

This is the way 😁

**Comment by 2bndy5** [comment]
My main concern is the refactor of public API members (`PyTypedField`) into  private modules (`sphinx.domains.python._object`). While these re-exports deal with backward compatibility, they might be considered "unused imports" at some point in the future, thus a possible subject for removal in 2 major versions.

Many linters have warnings about importing from private modules, so any third-party extension that might want to implement their own autodoc behavior would have to either
- hope what they need has been re-exported via public modules
- ignore the linter warning and import from private modules

Simply put, it would be more PEP8 compliant if the any refactored modules that contain previously public API to be renamed as public modules. Change `sphinx.domains.python._object` -> `sphinx.domains.python.object`.

**Comment by chrisjsewell** [comment]
> Simply put, it would be more PEP8 compliant if the any refactored modules that contain previously public API to be renamed as public modules.

Not that's not necessary,  you just use `__all__`, as discussed here: https://github.com/sphinx-doc/sphinx/issues/12246

**Comment by 2bndy5** [comment]
Using `__all__` is certainly a solution, but maintaining that is additional tech debt. that said it might be a good solution if one of the new private modules contains both public and private API.
